pkg/registry/core/service/ipallocator/controller/repair.go

```diff
@@ -313,3 +331,69 @@ func (c *Repair) checkLeaked(leaks map[string]int, stored ipallocator.Interface,
 		}
 	})
 }
+
+// used to generate empty patch bytes
+type emptyType struct {
+}
+
+// Services with "PreferDualStack" would be
+// -- set to dual-stack on a dual-stack cluster.
+// -- set to single stack on a single stack cluster.
+// Because the way allocator work (on Create/Update) they won't be upgraded
+// downgraded (from/to dual-stack) as cluster config change. More over
+// they will be arbitrary upgraded/downgraded if the user tried to
+// update them (e.g. setting a label).
+// This func ensures that all services of policy "PreferDualStack" is aligned
+// to cluster config without the user needing to update them
+func (c *Repair) fixPreferDualStackServices() error {
+	isClusterDualStack := len(c.allocatorByFamily) == 2
+
+	svcList, err := c.serviceClient.Services(metav1.NamespaceAll).List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return fmt.Errorf("preferDualStackFixer: unable to get list of services with err: %v", err)
+	}
+
+	for _, svc := range svcList.Items {
+		if svc.Spec.Type != v1.ServiceTypeClusterIP ||
+			svc.Spec.IPFamilyPolicy == nil ||
+			svc.Spec.ClusterIP == v1.ClusterIPNone {
+			// not one of the types we operate on
+			continue
+		}
+
+		if *(svc.Spec.IPFamilyPolicy) == v1.IPFamilyPolicyPreferDualStack {
+			bShouldChange := false
+			title := ""
+			message := ""
+			// should upgrade or dwongrade
+			if len(svc.Spec.ClusterIPs) == 1 && isClusterDualStack {
+				bShouldChange = true
+				title = "DualStackUpgrade"
+				message = fmt.Sprintf("service %ve/%v with policy:PreferDualStack has a single assigned ClusterIP. The service was upgraded", svc.Namespace, svc.Name)
+			}
+			if len(svc.Spec.ClusterIPs) == 2 && !isClusterDualStack {
+				bShouldChange = true
+				title = "DualStackDowngrade"
+				message = fmt.Sprintf("service %ve/%v with policy:PreferDualStack has two assigned ClusterIPs. The service was downgraded", svc.Namespace, svc.Name)
+
+			}
+
+			if bShouldChange {
+				// generate empty patch bytes
+				empty := emptyType{}
+				patchBytes, err := json.Marshal(&empty)
+				if err != nil {
+					return fmt.Errorf("preferDualStackFixer: failed to generate patch bytes for service %v/%v with err:%v", svc.Namespace, svc.Name, err)
+				}
+
+				if _, err = c.serviceClient.Services(svc.Namespace).Patch(context.TODO(), svc.Name, types.StrategicMergePatchType, patchBytes, metav1.PatchOptions{}); err != nil {
+					return fmt.Errorf("preferDualStackFixer: failed to patch service:%v/%v with err:%v", svc.Namespace, svc.Name, err)
```

Comment:
should we fail hard on the first error, or accumulate or just log?

Reply:
Any error here should be transient. and should be corrected in next run. Both choices are correct. Let me think this over.

Reply:
it is now broadcasting events instead of failure.